test/integration/transport_socket_matcher_integration_test.cc

```diff
@@ -0,0 +1,444 @@
+    setUpstreamProtocol(Http::CodecType::HTTP1);
+  }
+
+  enum class MatchType { EndpointMetadata };
+
+protected:
+  const uint32_t num_hosts_{2};
+  const std::string host_type_header_{"x-host-type"};
+  const std::string host_header_{"x-host"};
+  const std::string type_key_{"type"};
+  bool use_matcher_{true};
+  MatchType match_type_{MatchType::EndpointMetadata};
+
+  Http::TestRequestHeaderMapImpl tls_request_headers_{{":method", "GET"},
+                                                      {":path", "/test"},
+                                                      {":scheme", "http"},
+                                                      {":authority", "host"},
+                                                      {"x-route-type", "tls"}};
+  Http::TestRequestHeaderMapImpl raw_request_headers_{{":method", "GET"},
+                                                      {":path", "/test"},
+                                                      {":scheme", "http"},
+                                                      {":authority", "host"},
+                                                      {"x-route-type", "raw"}};
+};
+
+// Test the xDS-based transport socket matcher end-to-end using endpoint metadata input.
+TEST_F(TransportSocketMatcherIntegrationTest, XdsMatcherEndpointMetadata) {
+  initialize();
+
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  // Requests should route to the correct upstream sockets based on endpoint metadata type.
+  for (int i = 0; i < 3; i++) {
+    IntegrationStreamDecoderPtr response =
+        codec_client_->makeHeaderOnlyRequest(tls_request_headers_);
+    ASSERT_TRUE(response->waitForEndStream());
+    EXPECT_EQ("200", response->headers().getStatusValue());
+    const auto header_entry = response->headers().get(Http::LowerCaseString{host_type_header_});
+    EXPECT_FALSE(header_entry.empty());
+    EXPECT_EQ("tls", header_entry[0]->value().getStringView());
+
+    response = codec_client_->makeHeaderOnlyRequest(raw_request_headers_);
+    ASSERT_TRUE(response->waitForEndStream());
+    EXPECT_EQ("200", response->headers().getStatusValue());
+    const auto header_entry2 = response->headers().get(Http::LowerCaseString{host_type_header_});
+    EXPECT_FALSE(header_entry2.empty());
+    EXPECT_EQ("raw", header_entry2[0]->value().getStringView());
+  }
+}
+
+// Test xDS matcher using SourceIP input with prefix matching on loopback.
+TEST_F(TransportSocketMatcherIntegrationTest, XdsMatcherSourceIP) {
+  // Override matcher to use source_ip input; map IPv4 loopback prefix to TLS, fallback to raw.
+  config_helper_.addConfigModifier([](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+    auto* cluster = bootstrap.mutable_static_resources()->mutable_clusters(0);
+
+    xds::type::matcher::v3::Matcher matcher;
+    const std::string matcher_yaml = R"EOF(
+matcher_list:
+  matchers:
+  - predicate:
+      single_predicate:
+        input:
+          name: envoy.matching.inputs.source_ip
```

Comment:
This test needs to be updated I think?

Reply:
These are not needed anymore. I'll delete them.